From 4f3098ced691b7c7fa1f5bde8bf33bece5d25197 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Fri, 16 Sep 2022 15:10:47 +0200
Subject: [PATCH 0551/1695] ta: remoteproc: add the platform TLV support

The platform TLV chunk is parsed and each TLV is sent to the platform
PTA thanks to the PTA_REMOTEPROC_TLV_PARAM service.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0bacd30639c06d074f52fc1d76cf82c41a14a5fe ta: remoteproc: add the platform TLV support]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/266710
Change-Id: Ia8b7a8b34f7e108a06c13dcbe00f8098dc23bf41
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/348405
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 ta/remoteproc/src/remoteproc_core.c | 47 +++++++++++++++++++++++++++++
 1 file changed, 47 insertions(+)

diff --git a/ta/remoteproc/src/remoteproc_core.c b/ta/remoteproc/src/remoteproc_core.c
index c3b815d9b..ce958dff1 100644
--- a/ta/remoteproc/src/remoteproc_core.c
+++ b/ta/remoteproc/src/remoteproc_core.c
@@ -483,6 +483,48 @@ static TEE_Result get_rproc_pta_capabilities(struct remoteproc_context *ctx)
 	return TEE_SUCCESS;
 }
 
+static TEE_Result remoteproc_set_platform_tlv(struct remoteproc_context *ctx)
+{
+	const uint32_t  exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						 TEE_PARAM_TYPE_VALUE_INPUT,
+						 TEE_PARAM_TYPE_MEMREF_INPUT,
+						 TEE_PARAM_TYPE_NONE);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	uint8_t *p_tlv = ctx->tlvs;
+	uint8_t *p_end_tlv = p_tlv + ctx->tlvs_sz;
+	uint16_t tlv_length = 0;
+	uint32_t tlv_type = 0;
+	uint32_t tlv_v = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	params[0].value.a = ctx->rproc_id;
+
+	/* Parse the tlv area */
+	while (p_tlv < p_end_tlv) {
+		memcpy(&tlv_v, p_tlv, sizeof(tlv_v));
+		tlv_type = TEE_U32_FROM_LITTLE_ENDIAN(tlv_v);
+		memcpy(&tlv_v, p_tlv + RPROC_TLV_LENGTH_OF, sizeof(tlv_v));
+		tlv_length = TEE_U32_FROM_LITTLE_ENDIAN(tlv_v);
+
+		if (tlv_type >= RPROC_PLAT_TLV_TYPE_MIN &&
+		    tlv_type < RPROC_PLAT_TLV_TYPE_MAX) {
+			params[1].value.a  = tlv_type;
+			params[2].memref.size = tlv_length;
+			params[2].memref.buffer = &p_tlv[RPROC_TLV_VALUE_OF];
+
+			res = TEE_InvokeTACommand(pta_session,
+						  TEE_TIMEOUT_INFINITE,
+						  PTA_REMOTEPROC_TLV_PARAM,
+						  exp_pt, params, NULL);
+			if (res)
+				return res;
+		}
+		p_tlv += ROUNDUP_64(sizeof(struct remoteproc_tlv) + tlv_length);
+	}
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result remoteproc_verify_firmware(struct remoteproc_context *ctx,
 					     uint8_t *fw_orig,
 					     uint32_t fw_orig_size)
@@ -520,6 +562,11 @@ static TEE_Result remoteproc_verify_firmware(struct remoteproc_context *ctx,
 	if (res)
 		goto free_sec_cpy;
 
+	/* TODO:  transmit to the PTA the platform tlv chunk */
+	res = remoteproc_set_platform_tlv(ctx);
+	if (res)
+		goto free_sec_cpy;
+
 	/* Store location of the loadable binary in non-secure memory */
 	ctx->fw_img_sz = hdr->img_len;
 	ctx->fw_img = FW_IMG_PTR(fw_orig, hdr);
-- 
2.47.2

